language-plutus-core/src/Language/PlutusCore/Evaluation/Machine/Cek.hs

```diff
@@ -77,45 +68,116 @@ import           Data.HashMap.Monoidal
 import qualified Data.Map                                           as Map
 import           Data.Text.Prettyprint.Doc
 
+import           Data.Array
+
 {- Note [Scoping]
-The CEK machine does not rely on the global uniqueness condition, so the renamer pass is not a
+   The CEK machine does not rely on the global uniqueness condition, so the renamer pass is not a
 prerequisite. The CEK machine correctly handles name shadowing.
 -}
 
+type TermWithMem uni = WithMemory Term uni
+type TypeWithMem uni = Type TyName uni ExMemory
+type KindWithMem = Kind ExMemory
+
+-- 'Values' for the modified CEK machine.
+data CekVal uni =
+    -- TODO: we probably want to store a @Some (ValueOf uni)@ here, but then we have trouble in
+    -- 'readKnownCek'. I'll reconsider the way we deal with annotations once again.
+    VCon (TermWithMem uni)
+  | VTyAbs ExMemory TyName KindWithMem (TermWithMem uni) (CekValEnv uni)
```

Comment:
> I'm rusty with how the resource tracking works, but do we need memory annotations on the values we return? I think the answer is "yes, because we need to look at them to compute costs for builtin applications".

Yes, `applyEvaluate` takes two values and applies one to another and when the argument is a final argument of a builtin, we need to traverse the list of arguments and compute the cost of the operation we're about to perform, so those arguments have to be somehow annotated and putting `ExMemory` directly into the AST seems like a very sensible option (we could think about providing it separately, but given that we reconstruct a term when we return a final result from the machine, I think it makes sense to be as close to `Term` as possible and that means putting annotations into the AST of `CekVal`).